# Genetic drift, not life history or RNAi, determine long term evolution of transposable elements
## Amir Szitenberg, Soyeon Cha, Charles H. Opperman, David M. Bird, Mark L. Blaxter, David H. Lunt

Evolutionary Biology Group, School of Environmental Sciences, University of Hull, Hull, UK   
Department of Plant Pathology, North Carolina State University, Raleigh, NC, USA  
Institute of Evolutionary biology, School of Biological Sciences, University of Edinburgh, Edinburgh, UK  

Corresponding author: E-mail: szitenberg@gmail.com  

<p xmlns:dct="http://purl.org/dc/terms/" xmlns:vcard="http://www.w3.org/2001/vcard-rdf/3.0#">
  <a rel="license"
     href="http://creativecommons.org/publicdomain/zero/1.0/">
    <img src="http://i.creativecommons.org/p/zero/1.0/88x31.png" style="border-style: none;" alt="CC0" />
  </a>
</p>
  
[![DOI](https://zenodo.org/badge/22650/HullUni-bioinformatics/Nematoda-TE-Evolution.svg)](https://zenodo.org/badge/latestdoi/22650/HullUni-bioinformatics/Nematoda-TE-Evolution)  
   
###[Preprint](http://biorxiv.org/content/early/2015/12/20/034884.abstract)  
   
###[Supplementary information](https://dx.doi.org/10.6084/m9.figshare.c.3270356.v1)  

## Jupyter Notebooks

1. [Build Nematoda 18S rRNA tree, calculate assembly GC content & length](https://github.com/HullUni-bioinformatics/Nematoda-TE-Evolution/blob/master/18S_tree.ipynb)  
2. [Make denovo TE libraries](https://github.com/HullUni-bioinformatics/Nematoda-TE-Evolution/blob/master/Example_RepeatModeler_Run.ipynb)  
3. [Make a non-redundant Nematoda TE library](https://github.com/HullUni-bioinformatics/Nematoda-TE-Evolution/blob/master/non_redundant_Nematoda_TE_lib.ipynb)  
4. [DNA sequence homology TE search](https://github.com/HullUni-bioinformatics/Nematoda-TE-Evolution/blob/master/RepeatMasker.ipynb)    
5. [Assemble RepeatMasker detected TEs](https://github.com/HullUni-bioinformatics/Nematoda-TE-Evolution/blob/master/OneCodeToFindThemAll.ipynb)    
6. [Alternative TE detection approaches](https://github.com/HullUni-bioinformatics/Nematoda-TE-Evolution/blob/master/TransposonPSI_and_LTRHarvest.ipynb)  
7. [Merge TEs detected by different tools](https://github.com/HullUni-bioinformatics/Nematoda-TE-Evolution/blob/master/Integrated_TE_search_results.ipynb)  
8. [RNAi proteins honology search](https://github.com/HullUni-bioinformatics/Nematoda-TE-Evolution/blob/master/8_Search_RNAi_TE_silencing_pathway_proteins.ipynb)   
9. [Figures 1, 2 and S1](https://github.com/HullUni-bioinformatics/Nematoda-TE-Evolution/blob/master/9_Figure1%262%26S1_No_WTSI_genomes.ipynb)   
10. [Evolutionary analysis (Figs3, 4, 5, S2, S3, S4)](https://github.com/HullUni-bioinformatics/Nematoda-TE-Evolution/blob/master/10_EvolutionaryAnalysis.ipynb)  
11. [Test inverse corrolation between TE median sequence length and loads](https://github.com/HullUni-bioinformatics/Nematoda-TE-Evolution/blob/master/11_Ectopic_recombination_signature.ipynb)  